PCL usage requirements are not properly propagated by YARP_pcl

**Background:**
I have been recently working on a YARP wrapper around several PCL cloud processing and surface reconstruction algorithms ([library](https://github.com/roboticslab-uc3m/vision/tree/67df1286484e8566c238fe750a522a6f4b41bce5/libraries/YarpCloudUtils), [rationale](https://github.com/roboticslab-uc3m/vision/issues/102)). The YARP_pcl component had come in useful in terms of YARP2PCL and PCL2YARP type conversions, although not devoid of some shortcomings on the CMake side of life.

**Minimal not working example:**

```cmake
# CMakeLists.txt
cmake_minimum_required(VERSION 3.12 FATAL_ERROR)
project(yarp-pcl LANGUAGES CXX)
find_package(YARP 3.4 REQUIRED pcl)
add_executable(${PROJECT_NAME} main.cpp)
target_link_libraries(${PROJECT_NAME} YARP::YARP_pcl)
```

```cxx
// main.cpp
#include <yarp/pcl/Pcl.h>
auto main() -> int {}
```

```
In file included from /home/bartek/Desktop/temp/yarp-pcl/main.cpp:1:
/usr/local/include/yarp/pcl/Pcl.h:12:10: fatal error: pcl/io/pcd_io.h: No such file or directory
   12 | #include <pcl/io/pcd_io.h>
      |          ^~~~~~~~~~~~~~~~~
```

**Diagnosis:**
As expected, YARP_pcl is instructed to propagate PCL usage requirements to downstreams:

https://github.com/robotology/yarp/blob/0481f994c6e03897d038c5f1d1078145646a1772/src/libYARP_pcl/src/CMakeLists.txt#L23-L25

However, the `${PCL_INCLUDE_DIR}` and `${PCL_LIBRARIES}` variables are **empty**. These have not been populated since YARP does not perform a `find_package(PCL)` call anywhere. As a result, downstream projects are instructed to search for PCL on the system (thanks to the third instruction), but neither PCL targets/libraries nor PCL include directories are automatically propagated via `target_link_libraries(... YARP::YARP_pcl)` as they should be, thus resulting in the missing header inclusion error.

**Remark regarding YARP_cv:**
I believe this component was developed in a similar vein to YARP_cv. Please note **YARP_cv does not export OpenCV's usage requirements either**. Even though I haven't tested it thoroughly, I believe this same issue prevails after inspecting the generated *YARP_cvTargets.cmake* file. In that case, however, I tracked down the problem to the fact that there is no such thing as `OPENCV_INCLUDE_DIR` ([ref](https://github.com/robotology/yarp/blob/0481f994c6e03897d038c5f1d1078145646a1772/src/libYARP_cv/src/CMakeLists.txt#L24-L25)) on my OpenCV 4.4 installation, it should read `OpenCV_INCLUDE_DIRS` instead or, ideally, favor modern CMake targets.

Besides, YARP actually looks for OpenCV as there exists some CV-dependent code such as the *opencv_grabber* device.

**Additional suggestions:**
There is more room for improvements beyond fixing the `find_package` thing:
* In [yarp/pcl/Pcl.h](https://github.com/robotology/yarp/blob/0481f994c6e03897d038c5f1d1078145646a1772/src/libYARP_pcl/src/yarp/pcl/Pcl.h), I'd add these header inclusions:
  ```cxx
  #include <cstring> // std::memcpy, also mind prepending std:: to the existing call
  #include <pcl/point_cloud.h> // pcl::PointCloud<T>
  ```
* The following line in [CMakeLists.txt](https://github.com/robotology/yarp/blob/0481f994c6e03897d038c5f1d1078145646a1772/src/libYARP_pcl/src/CMakeLists.txt):
  ```cmake
  list(APPEND YARP_pcl_PUBLIC_DEPS PCL)
  ```
  is better written as:
  ```cmake
  list(APPEND YARP_pcl_PUBLIC_DEPS "PCL COMPONENTS common io")
  ```
  In this way, downstreams only need to search for specific PCL components as requested by YARP_pcl. Otherwise, *all* available PCL components and their related variables are needlessly populated and brought to scope.
* Care should be taken to check this against different PCL versions, but on PCL 1.10 I can confirm `PCL_INCLUDE_DIR` is empty. It should read `PCL_INCLUDE_DIRS` instead.

**Tested configuration:**
* YARP version 3.4.1+81-20201214.6+git1b0060380
* PCL 1.10 (via `apt install libpcl-dev`)
* cmake version 3.16.3
* g++ (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
* Ubuntu 20.04.1 (focal), Linux 5.4.0-58-generic

--- a/cmake/YarpFindDependencies.cmake
+++ b/cmake/YarpFindDependencies.cmake
@@ -334,6 +334,9 @@ checkandset_dependency(OpenCV)
 set_property(DIRECTORY PROPERTY INCLUDE_DIRECTORIES ${OpenCV24_WORKAROUND_ORIGINAL_INCLUDE_DIRS})
 # End second part of workaround
 
+find_package(PCL COMPONENTS common io QUIET)
+checkandset_dependency(PCL)
+
 find_package(Lua QUIET)
 checkandset_dependency(Lua)
 
@@ -608,6 +611,7 @@ print_dependency(QGVCore)
 print_dependency(Libedit)
 print_dependency(SWIG)
 print_dependency(OpenCV)
+print_dependency(PCL)
 print_dependency(Lua)
 print_dependency(LibOVR)
 print_dependency(GLFW3)

--- a/doc/release/yarp_3_4/fix_interface_lib_deps.md
+++ b/doc/release/yarp_3_4/fix_interface_lib_deps.md
@@ -1,0 +1,9 @@
+fix_interface_lib_deps {#yarp_3_4}
+----------------------
+
+## Build System
+
+* Dependencies for the `INTERFACE` libraries (YARP_cv, YARP_pcl)
+  are now properly set (#2438).
+* PCL is now required to build the YARP_pcl library.
+* OpenCV is now required to build the YARP_cv library.

--- a/src/libYARP_cv/CMakeLists.txt
+++ b/src/libYARP_cv/CMakeLists.txt
@@ -6,6 +6,13 @@
 
 project(YARP_cv)
 
+# OpenCV is required, otherwise the OpenCV_INCLUDE_DIR and OpenCV_LIBRARIES
+# (used in the CMakeLists.txt file in the src direcrory) are not expanded).
+# FIXME Remove this check when OpenCV targets will be used.
+if (NOT YARP_HAS_OpenCV)
+  return()
+endif()
+
 add_subdirectory(src)
 
 include(YarpInstallBasicPackageFiles)

--- a/src/libYARP_cv/src/CMakeLists.txt
+++ b/src/libYARP_cv/src/CMakeLists.txt
@@ -21,8 +21,8 @@ target_compile_features(YARP_conf INTERFACE cxx_std_14)
 target_link_libraries(YARP_cv INTERFACE YARP::YARP_sig)
 list(APPEND YARP_cv_PUBLIC_DEPS YARP_sig)
 
-target_include_directories(YARP_cv SYSTEM INTERFACE ${OPENCV_INCLUDE_DIR})
-target_link_libraries(YARP_cv INTERFACE ${OPENCV_INCLUDE_DIR})
+target_include_directories(YARP_cv SYSTEM INTERFACE ${OpenCV_INCLUDE_DIR})
+target_link_libraries(YARP_cv INTERFACE ${OpenCV_LIBRARIES})
 list(APPEND YARP_cv_PUBLIC_DEPS OpenCV)
 
 # set_property(TARGET YARP_cv PROPERTY PUBLIC_HEADER ${YARP_cv_HDRS})

--- a/src/libYARP_pcl/CMakeLists.txt
+++ b/src/libYARP_pcl/CMakeLists.txt
@@ -6,6 +6,14 @@
 
 project(YARP_pcl)
 
+# PCL is required, otherwise the PCL_<component>_INCLUDE_DIR and
+# PCL_<component>_LIBRARIES
+# (used in the CMakeLists.txt file in the src direcrory) are not expanded).
+# FIXME Remove this check when PCL targets will be used.
+if (NOT YARP_HAS_PCL)
+  return()
+endif()
+
 add_subdirectory(src)
 
 include(YarpInstallBasicPackageFiles)

--- a/src/libYARP_pcl/src/CMakeLists.txt
+++ b/src/libYARP_pcl/src/CMakeLists.txt
@@ -20,9 +20,11 @@ target_compile_features(YARP_pcl INTERFACE cxx_std_14)
 target_link_libraries(YARP_pcl INTERFACE YARP::YARP_sig)
 list(APPEND YARP_pcl_PUBLIC_DEPS YARP_sig)
 
-target_include_directories(YARP_pcl SYSTEM INTERFACE ${PCL_INCLUDE_DIR})
-target_link_libraries(YARP_pcl INTERFACE ${PCL_LIBRARIES})
-list(APPEND YARP_pcl_PUBLIC_DEPS PCL)
+target_include_directories(YARP_pcl SYSTEM INTERFACE ${PCL_COMMON_INCLUDE_DIR}
+                                                     ${PCL_IO_INCLUDE_DIR})
+target_link_libraries(YARP_pcl INTERFACE ${PCL_COMMON_LIBRARIES}
+                                         ${PCL_IO_LIBRARIES})
+list(APPEND YARP_pcl_PUBLIC_DEPS "PCL COMPONENTS common io")
 
 # set_property(TARGET YARP_pcl PROPERTY PUBLIC_HEADER ${YARP_pcl_HDRS})
 # set_property(TARGET YARP_pcl PROPERTY PRIVATE_HEADER ${YARP_pcl_IMPL_HDRS})

